### DISCUSSION TITLE
Solve failed when reading a mesh with assigning an individual grain ID for each grid

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#20800] [Q&A Modules: Solid mechanics] [answered]
Hi Moose experts or Users,

Recently, I developed a custom class inherited from `CrystalPlasticityStressUpdateBase` to study creep behavior. For the mesh with 1x1x1, the program can be smoothly performed according to the setting in the Executioner module. But for the mesh with 4x4x4 in which each grid is assigned an individual grain ID with the help of `ElementSubdomainIDGenerator`, the test file just can be run for a short time and then said "Solve failed and timestep already at or below dtmin, cannot continue!" I try to adjust the timestep but that problem still happened. Any ideas for this? Thank you.

**Comment by GiudGiud** [comment]
Hello

Could you please attach the whole output log?
And the input as well.

Guillaume

**Reply by xchengood** [reply]
Hi @GiudGiu, thank you for your reply. The output log, test file, and input_file used in `ElementPropertyReadFile` are here.
[output log.txt](https://github.com/idaholab/moose/files/8497995/output.log.txt)
[test.txt](https://github.com/idaholab/moose/files/8497997/test.txt)
[input_file.txt](https://github.com/idaholab/moose/files/8497998/input_file.txt)

**Reply by sapitts** [reply]
@xchengood, Please make sure that the slip systems and unit cell lengths are consistent with the crystal lattice type modeled in the crystal plasticity material class you have selected:
- You have used `CrystalPlasticityHCPDislocationSlipBeyerleinUpdate` as the material class in your input file, which assumes an HCP lattice (unless you have modified the code locally)
- In contrast, your input slip systems have the name for a BCC crystal, and the unit cell dimensions used in your input file are all equal, as is appropriate for a cubic crystal but not a HCP crystal

Beyond these potential issues, the recommendations from @ngrilli to check your Euler angles and your convergence criteria are steps I would also recommend

**Reply by xchengood** [reply]
Hi @sapitts. Thank you for your advice. Actually, I did modify the code locally and changed the crystal type to BCC. I will follow your and @ngrilli's advice to do more attempts.

**Reply by sapitts** [reply]
Oh good, I'm glad that you had locally modified the code. The original code should have had error checks that would prevent accidental mixing of crystal lattice types--I was a bit worried that those checks had failed!

**Comment by ngrilli** [comment, marked as answer]
@xchengood  I cannot see anything obviously wrong in your input file,
but my suggestion is to plot the Euler angles and see if they correspond to what you expect.

Second, my experience with the crystal plasticity model is that when you start adding
more elements and grains, the convergence becomes more difficult,
therefore I think the criterion: nl_rel_tol = 1e-10 may be too strict, just leave default.
Also, definitely allow more non-linear iterations.
Also, use the substepping option of the class ComputeMultipleCrystalPlasticityStress
and increase the number of iterations for stress and state variables.

Third, there seems to be a pressure boundary condition with a jump in stress,
don't do that, increase the pressure gradually from zero to the maximum value.
Also, try displacement control instead.

Best Regards,
Nicolò

**Reply by xchengood** [reply]
Hi @ngrilli. Thank you for your advice. I will follow up to do some modifications and do more attempts.
